http/oidc/src/main/java/org/wildfly/security/http/oidc/JWKEncPublicKeyLocator.java

```diff
@@ -0,0 +1,109 @@
+import static org.apache.http.HttpHeaders.ACCEPT;
+import static org.wildfly.security.http.oidc.ElytronMessages.log;
+import static org.wildfly.security.http.oidc.Oidc.JSON_CONTENT_TYPE;
+
+import java.security.PublicKey;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.apache.http.client.methods.HttpGet;
+import org.wildfly.security.jose.jwk.JWK;
+import org.wildfly.security.jose.jwk.JWKParser;
+import org.wildfly.security.jose.jwk.JsonWebKeySet;
+
+/**
+ * A public key locator that dynamically obtains the public key used for encryption
+ * from an OpenID provider by sending a request to the provider's {@code jwks_uri}
+ * when needed.
+ *
+ * @author <a href="mailto:prpaul@redhat.com">Prarthona Paul</a>
+ * */
+class JWKEncPublicKeyLocator implements PublicKeyLocator {
+    private Map<String, PublicKey> currentKeys = new ConcurrentHashMap<>();
+
+    private volatile int lastRequestTime = 0;
+
+    @Override
+    public PublicKey getPublicKey(String kid, OidcClientConfiguration config) {
+        int minTimeBetweenRequests = config.getMinTimeBetweenJwksRequests();
+        int publicKeyCacheTtl = config.getPublicKeyCacheTtl();
+        int currentTime = getCurrentTime();
+
+        PublicKey publicKey = lookupCachedKey(publicKeyCacheTtl, currentTime, JWK.Use.ENC.toString());
+        if (publicKey != null) {
+            return publicKey;
+        }
+
+        synchronized (this) {
+            currentTime = getCurrentTime();
+            if (currentTime > lastRequestTime + minTimeBetweenRequests) {
+                sendRequest(config);
+                lastRequestTime = currentTime;
+            } else {
+                log.debug("Won't send request to jwks url. Last request time was " + lastRequestTime);
+            }
+            return lookupCachedKey(publicKeyCacheTtl, currentTime, JWK.Use.ENC.toString());
+        }
+
+    }
+
+    @Override
+    public void reset(OidcClientConfiguration config) {
+        synchronized (this) {
+            sendRequest(config);
+            lastRequestTime = getCurrentTime();
+        }
+    }
+
+    private PublicKey lookupCachedKey(int publicKeyCacheTtl, int currentTime, String kid) {
+        if (lastRequestTime + publicKeyCacheTtl > currentTime && kid != null) {
+            return currentKeys.get(kid);
+        } else {
+            return null;
+        }
+    }
+
+    private static int getCurrentTime() {
+        return (int) (System.currentTimeMillis() / 1000);
+    }
+
+    private void sendRequest(OidcClientConfiguration config) {
+        if (log.isTraceEnabled()) {
+            log.trace("Going to send request to retrieve new set of public keys to encrypt a JWT request for client " + config.getResourceName());
+        }
+
+        HttpGet request = new HttpGet(config.getJwksUrl());
+        request.addHeader(ACCEPT, JSON_CONTENT_TYPE);
+        try {
+            JWK[] jwkList = Oidc.sendJsonHttpRequest(config, request, JsonWebKeySet.class).getKeys();
```

Comment:
Can we use JsonWebKeySetUtil.getKeysForUse here with JWK.Use.ENC? (like https://github.com/wildfly-security/wildfly-elytron/blob/2.x/http/oidc/src/main/java/org/wildfly/security/http/oidc/JWKPublicKeyLocator.java#L98)

Reply:
I dont think so, because then lookupCachedKey returns null. 
The existing code uses keyUse string as the key for the map, while JsonWebKeySetUtil.getKeysForUse uses the key id.

Reply:
Thanks for clarifying that! Just to check, any reason to not use the key id as the key for the map?

Reply:
This is because in this case keyid is not known until we have already extracted the keys.

Reply:
Just to check, is it known for the JWK.Use.SIG case? (Just trying to understand the reason for the difference between the two cases)

Reply:
I see the difference now. 
The JWKEncPublicKeyLocator class implements PublicKeyLocator and also allows for caching where possible. This class is built with JWKPublicKeyLocator.java class as an example. 

The reason why we cant implement caching in JsonWebKeySetUtil is because of this line that uses keyid as the key.
```
result.put(jwk.getKeyId(), parser.toPublicKey());
```

When the JWKPublicKeyLocator.getPublicKey() class is called, the keyid is sent as a parameter from the TokenValidator class. However, when `JWKEncPublicKeyLocator.getPublicKey()` is called, the keyid field does not receive a real key id, instead it received the string value of the key USE, which is enc